### ISSUE TITLE
TCPSocket hangs when RST is received on Windows

### ISSUE THREAD

**Issue by yamam** [jruby/jruby#6346]
**Environment Information**

jruby 9.2.12.0 (2.5.7) 2020-07-01 db01a49ba6 Java HotSpot(TM) Client VM 25.161-b12 on 1.8.0_161-b12 +jit [mswin32-i386]

**Expected Behavior**

server.rb
```ruby
require 'socket'
sock = TCPServer.new(nil, 12345).accept
while line = sock.gets
    p line
end
```

client.rb
```ruby
require 'socket'
s = TCPSocket.open('localhost', 12345)
s.puts("hello")
sleep 1
# send RST on close
linger = [1,0].pack('ii')
s.setsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER, linger)
s.close
```

$ ruby server.rb
"hello\n"
Traceback (most recent call last):
        1: from server.rb:3:in `<main>'
server.rb:3:in `gets': An existing connection was forcibly closed by the remote host. @ io_fillbuf - fd:4  (Errno::ECONNRESET)

$ ruby client.rb

**Actual Behavior**

$ c:\jruby-9.2.12.0\bin\jruby.exe server.rb
"hello\n"

The program hangs and CPU utilization is high.

**Comment by headius** [comment]
Raises the error on Darwin but message could be improved:

```
"hello\n"
Errno::ECONNRESET: Connection reset by peer - channel: org.jruby.util.io.ChannelFD@38364841
    gets at org/jruby/RubyIO.java:2318
  <main> at server.rb:3
```

The code in this area does not differ significantly on Windows, so this might be a behavioral difference at the JDK level.

**Comment by rdp** [comment]
Seems like it's generating exception internally jconsole main thread:

```

Stack trace: 
java.base@20.0.2/java.lang.Throwable.fillInStackTrace(Native Method)
java.base@20.0.2/java.lang.Throwable.fillInStackTrace(Throwable.java:820)
   - locked java.net.SocketException@5687ca66
java.base@20.0.2/java.lang.Throwable.<init>(Throwable.java:273)
java.base@20.0.2/java.lang.Exception.<init>(Exception.java:67)
java.base@20.0.2/java.io.IOException.<init>(IOException.java:57)
java.base@20.0.2/java.net.SocketException.<init>(SocketException.java:47)
java.base@20.0.2/sun.nio.ch.SocketChannelImpl.throwConnectionReset(SocketChannelImpl.java:401)
java.base@20.0.2/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:418)
app//org.jruby.util.io.PosixShim.read(PosixShim.java:158)
app//org.jruby.util.io.OpenFile$2.run(OpenFile.java:1333)
app//org.jruby.util.io.OpenFile$2.run(OpenFile.java:1319)
app//org.jruby.RubyThread.executeReadWrite(RubyThread.java:1792)
app//org.jruby.util.io.OpenFile.readInternal(OpenFile.java:1404)
app//org.jruby.util.io.OpenFile.fillbuf(OpenFile.java:1298)
app//org.jruby.util.io.OpenFile.getlineFast(OpenFile.java:1638)
app//org.jruby.RubyIO.getlineImpl(RubyIO.java:717)
app//org.jruby.RubyIO$3.getline(RubyIO.java:2545)
app//org.jruby.RubyIO$3.getline(RubyIO.java:2542)
app//org.jruby.util.io.Getline.getlineCall(Getline.java:144)
app//org.jruby.util.io.Getline.getlineCall(Getline.java:23)
app//org.jruby.RubyIO.gets(RubyIO.java:2521)
app//org.jruby.RubyIO$INVOKER$i$gets.call(RubyIO$INVOKER$i$gets.gen)
app//org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193)
server.invokeOther3:gets(server.rb:3)
server.RUBY$script(server.rb:3)
server.run(server.rb)
java.base@20.0.2/java.lang.invoke.LambdaForm$DMH/0x0000017f32c9d400.invokeStatic(LambdaForm$DMH)
java.base@20.0.2/java.lang.invoke.LambdaForm$MH/0x0000017f3315b400.invoke(LambdaForm$MH)
java.base@20.0.2/java.lang.invoke.LambdaForm$MH/0x0000017f33159c00.invoke(LambdaForm$MH)
java.base@20.0.2/java.lang.invoke.Invokers$Holder.invokeExact_MT(Invokers$Holder)
java.base@20.0.2/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
app//org.jruby.ir.Compiler$1.load(Compiler.java:114)
app//org.jruby.Ruby.runScript(Ruby.java:1276)
app//org.jruby.Ruby.runNormally(Ruby.java:1193)
app//org.jruby.Ruby.runNormally(Ruby.java:1175)
app//org.jruby.Ruby.runNormally(Ruby.java:1211)
app//org.jruby.Ruby.runFromMain(Ruby.java:989)
app//org.jruby.Main.doRunFromMain(Main.java:398)
app//org.jruby.Main.internalRun(Main.java:282)
app//org.jruby.Main.run(Main.java:227)
app//org.jruby.Main.main(Main.java:199)
```

Maybe getting swallowed?

**Comment by headius** [comment]
This should be fixed by #7967.
